src/Orleans/Streams/PersistentStreams/IStreamQueueBalancer.cs

```diff
@@ -8,27 +8,32 @@ namespace Orleans.Streams
     /// The stream queue balancer is responsible for load balancing queues across all other related queue balancers.  It
     /// notifies any listeners (<code>IStreamQueueBalanceListener</code>) of changes to the distribution of queues.
     /// </summary>
-    internal interface IStreamQueueBalancer
+    public interface IStreamQueueBalancer
     {
         /// <summary>
+        /// Initialize this instance
+        /// </summary>
+        /// <returns></returns>
+        Task Initialize();
+        /// <summary>
         /// Retrieves the latest queue distribution for this balancer.
         /// </summary>
         /// <returns>Queue allocated to this balancer.</returns>
-        IEnumerable<QueueId> GetMyQueues();
+        Task<IEnumerable<QueueId>> GetMyQueues();
```

Comment:
But if we make this synchronous, for example, how would the balancer get its updated queues in a rebalancing scenario? for example, one silo shut down, current balancer need to locate more queues on it from the async leasing manager. How would this work as a injected balancer? 

The whole pluggable queue balancer feature is very limited. You cannot plug in a very powerful balancer without changing the internal runtime.